### PR TITLE
Presenter: Allow presentation wide default colour scheme.

### PR DESCRIPTION
**Pull request by jpgleeson**
Also allow toggling between a black and white background.
Makes some progress towards the items on  #15767 for black screen, white screen, and templates.
PresentationColorScheme

### CODE REVIEW

File: Userland/Applications/Presenter/Presentation.cpp [hunk 1]
```diff
@@ -16,6 +16,7 @@ Presentation::Presentation(Gfx::IntSize normative_size, HashMap<DeprecatedString
     : m_normative_size(normative_size)
     , m_metadata(move(metadata))
 {
+    m_background_color = Color::White;
```
Comment (kleinesfilmroellchen) [inline Userland/Applications/Presenter/Presentation.cpp:19]: Put this in the initializer list of the constructor

File: Userland/Applications/Presenter/Presentation.cpp [hunk 2]
```diff
@@ -42,6 +43,28 @@ StringView Presentation::author() const
...
+Color Presentation::foreground_color() const
+{
+    if (m_metadata.contains("foreground-color"sv)) {
+        auto color = Color::from_string(m_metadata.get("foreground-color"sv)->view());
+        if (color.has_value()) {
+            return color.value();
+        }
+    }
+    return Color::DarkGray;
+}
```
Comment (kleinesfilmroellchen) [inline Userland/Applications/Presenter/Presentation.cpp:66]: Because we need them regularly, they can be extracted and defaulted at creation time. This kind of extraction is only done for less important data, such as the modification date.

File: Userland/Applications/Presenter/Presentation.h [hunk 3]
```diff
@@ -37,6 +39,7 @@ class Presentation {
     void next_frame();
     void previous_frame();
     void go_to_first_slide();
+    void set_color_scheme(Gfx::Color, Gfx::Color);
```
Comment (kleinesfilmroellchen) [inline Userland/Applications/Presenter/Presentation.h:42]: Please mention the parameter names here as well; we only leave out parameter names for declarations if their purpose is obvious (e.g. setters).

File: Userland/Applications/Presenter/PresenterWidget.cpp [hunk 4]
```diff
@@ -68,6 +68,23 @@ ErrorOr<void> PresenterWidget::initialize_menubar()
...
+    auto white_background_action = GUI::Action::create("&White background", { KeyCode::Key_W }, [this](auto&) {
+        if (m_current_presentation) {
+            m_current_presentation->set_color_scheme(Color::White, Color::DarkGray);
+            update();
+        }
+    });
+    TRY(presentation_menu.try_add_action(black_background_action));
+    TRY(presentation_menu.try_add_action(white_background_action));
+    m_black_background_action = black_background_action;
+    m_white_background_action = white_background_action;
```
Comment (kleinesfilmroellchen) [inline Userland/Applications/Presenter/PresenterWidget.cpp:86]: This belongs in a separate commit; it adds the black-out and white-out functionality which relates to color scheme but is not the same (it's just mentioned in the documentation alongside).

File: Userland/Applications/Presenter/PresenterWidget.cpp [hunk 5]
```diff
@@ -79,6 +96,7 @@ void PresenterWidget::set_file(StringView file_name)
     } else {
         m_current_presentation = presentation.release_value();
         window()->set_title(DeprecatedString::formatted(title_template, m_current_presentation->title(), m_current_presentation->author()));
+        m_current_presentation->set_color_scheme(m_current_presentation->background_color(), m_current_presentation->foreground_color());
```
Comment (kleinesfilmroellchen) [inline Userland/Applications/Presenter/PresenterWidget.cpp:99]: That makes no sense at all; the presentation should handle its color scheme internally as it can load it while it is being created.

File: Userland/Applications/Presenter/SlideObject.cpp [hunk 6]
```diff
@@ -82,6 +82,16 @@ Text::Text()
...
+Text::Text(Gfx::Color default_color)
+{
+    REGISTER_STRING_PROPERTY("text", text, set_text);
+    REGISTER_FONT_WEIGHT_PROPERTY("font-weight", font_weight, set_font_weight);
+    REGISTER_TEXT_ALIGNMENT_PROPERTY("text-alignment", text_alignment, set_text_alignment);
+    REGISTER_INT_PROPERTY("font-size", font_size, set_font_size);
+    REGISTER_STRING_PROPERTY("font", font, set_font);
+    this->set_color(default_color);
+}
+
```
Comment (kleinesfilmroellchen) [inline Userland/Applications/Presenter/SlideObject.cpp:94]: You can remove the no-argument constructor of Text because the SlideObject parser function should be the only user of the SlideObject (and children) constructors.

File: Userland/Applications/Presenter/SlideObject.cpp [hunk 7]
```diff
@@ -82,6 +82,16 @@ Text::Text()
...
 }
 
+Text::Text(Gfx::Color default_color)
+{
+    REGISTER_STRING_PROPERTY("text", text, set_text);
+    REGISTER_FONT_WEIGHT_PROPERTY("font-weight", font_weight, set_font_weight);
+    REGISTER_TEXT_ALIGNMENT_PROPERTY("text-alignment", text_alignment, set_text_alignment);
+    REGISTER_INT_PROPERTY("font-size", font_size, set_font_size);
+    REGISTER_STRING_PROPERTY("font", font, set_font);
+    this->set_color(default_color);
```
Comment (kleinesfilmroellchen) [inline Userland/Applications/Presenter/SlideObject.cpp:92]: This can go in the initializer list of the constructor.

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]
So what you're implementing here is two new metadata keys: `foreground-color` and `background-color`. While that is convenient atm it's not very scalable; what if the colour schemes had more than just two colours? I still think my original idea, mostly based on named colour schemes, works better. We can relax the requirement for metadata entries to be strings, and allow colour scheme objects. So you can either have:

```json
"metadata": {
  "color-scheme": "light" // or "dark", "system", "Durrque", "Default", "Redmond", etc.
}
```

or

```json
"metadata": {
  "color-scheme": {
    "background-color": "#00ff00",
    "foreground-color": "#77ff00"
  }
}
```

It's totally fine if you only implement the first or the second here; however in any case please move the color scheme handling to Presentation construction. And also please document whatever you change about the format. You don't need to bump the version number as this is fully compatible if white is the default background color.

**Comment by jpgleeson** [comment]
https://github.com/SerenityOS/serenity/pull/16935 has made this change shape before I could revisit it. I'm closing this for now.
